### ISSUE TITLE
Global settings remain bolded even when changed back to their original setting

### ISSUE THREAD

**Issue by oddMLan** [project64/project64#653]
Steps to reproduce:
-Open the Settings
-Change a setting (like, changing a plugin, checking/unchecking Audio HLE or Graphics HLE)
-Click "Apply"
-Change back the changed settings to the default (don't use the "Reset Page" button, do it manually)
-Click "Apply"
-Setting still remains bolded, even thought it's the default setting.

Expected behaviour:
Setting should be unbolded when restored to its original value.

Restarting the emulator doesn't help; settings still remain bolded.
Only way to fix is to use the "Reset Page" or "Reset All" buttons.

Is this the intended behavior? Cause this is present since the Project64 1.7 builds.

**Comment by project64** [comment]
it is bold if there is some setting in the .cfg file overriding the default, even if it is the same as the default.

some ways like reset page, clear the setting out of the cfg file which is why it gets unbold, so the question is is it a bug in deleting the setting from the cfg fie.

**Comment by oddMLan** [comment]
So the setting should get deleted from the CFG when restored to its original value?

**Comment by cxd4** [comment]
Logically that's the only way for pj to remember whether you touched it at all.

Unless you were thinking add a second config file/store in the system registry a true/false flag signaling whether a setting was touched from its default value.  But the easiest is just seeing if it exists in a per-ROM cfg.

**Comment by oddMLan** [comment]
Should it matter to know if you touched that setting, when you set it back to the default value anyway?

My approach would be deleting the redundant setting from the CFG when it matches with the default in the RDB. So it doesn't get bolded for, no good reason other than pointing that you once touched that setting. In my opinion, the setting should only get bolded when the plugins are different from the default.

Unless again, this was the intended behavior since its implementation. In that case I think I can close the bug.

**Comment by AmbientMalice** [comment]
From a usability point of view, if the current setting matches the default, it should not be bolded. That's unnecessarily confusing. The entire *point* of the bolding, IMO, is to indicate "this no longer reflects the default".

**Comment by cxd4** [comment]
>My approach would be deleting the redundant setting from the CFG when it matches with the default in the RDB.

That's what I was trying to say.
There's no other logical approach to doing what you mean to be done.

Actually, I haven't been able to reproduce this issue ever since Project64 1.7 like you said.

Every time I set stuff like RDB settings through the GUI, then back to what they were before and click "Apply" then it gets un-bolded.  So I haven't seen your issue.

**Comment by cxd4** [comment]
Actually, I haven't been able to reproduce this issue ever since Project64 1.7 like you said.

Every time I set stuff like RDB settings through the GUI, then back to what they were before and click "Apply" then it gets un-bolded.  So I haven't seen your issue.

**Comment by cxd4** [comment]
Anybody else can reproduce this issue?  Because so far I can't.

The only time I remember seeing this issue happen is when changing global plugin settings, and that's sort of understandable since we have no rigid concept of default plugins anyway.  (Per-ROM plugins aren't bolded like this issue suggests though.)

**Comment by oddMLan** [comment]
You're right, this doesn't affect Per-ROM settings.
I was really talking about the Global settings.

**Comment by Frank-74** [comment]
This is still a bit annoying with for example this dialog:
![boldsettings](https://cloud.githubusercontent.com/assets/11056843/20609031/8929c2fc-b27f-11e6-9bc6-d3ba65e05a6c.png)

Because I once enabled debugger, it remains bold forever, unless I reset all the settings.  Clicking Apply should un-bold it.

**Comment by project64** [comment]
if there is a setting in the .cfg file it means it overrides the default setting (even if it overrides it with the same setting) It which case it should be bold. So if you check the cfg file do you see it there.

Reset deletes all the settings, so should be removed there.

Setting back to the original value, there is code there that tries to detect it matches the default in which case delete it out of the cfg file and use the original default. 

So it should only be unbolded if it is not in the .cfg file

**Comment by Frank-74** [comment]
[debugger] section doesn't exist in a fresh Project64.cfg.  [debugger] section is only written once it is first enabled.  So it will always be bold unless I clear the settings.

**Comment by oddMLan** [comment]
>if there is a setting in the .cfg file it means it overrides the default setting (even if it overrides it with the same setting) It which case it should be bold. 

But what is that useful for? It's unnecesarily confusing. If it matches the default, the redundant should be removed from the cfg

**Comment by project64** [comment]
it is basically saying using an override or the default .. while the override maybe the same and so it looks confusing it is different and hence while bold. But because it is confusing I added the code to delete the override if it is set back to the default. 

So that it is bold is not an issue, the issue is why does it not delete the override if set back to a default setting.

**Comment by oddMLan** [comment]
Actually the code that removes the redundant entries works in the per-game settings panel, but not in the global settings

**Comment by cxd4** [comment]
I'm guessing that's because the per-game settings are the override settings, while the global settings are system defaults and by definition are therefore never redundant.

**Comment by oddMLan** [comment]
Then how does the emulator know when a global setting doesn't match the default, and therefore, bolded? It's just a matter of applying the same principle to the global settings.

**Comment by cxd4** [comment]
Global settings are the default settings...?  If they weren't then how could they be global.

I was very annoyed by the bolded settings names since Project64 1.7 in 2008 or so too.  I don't know why but it's 8 years later and I see it as logical now for some reason.  It's basically preferential but nothing really wrong with it either.  You think there's something wrong with it because the reality of configuration storage isn't obvious to most users.  Probably this was what zilmar was trying to tell me a decade ago but he was just very bad at explaining it with effective grammar and spelling or something so I never got it.

**Comment by oddMLan** [comment]
AFAIK the true "defacto" settings are coded in the executable. When a setting gets changed, it gets written to the Project64.cfg. There is a missing comparison between the settings coded in the executable and the settings found in the cfg file, so even if the default setting coded in the Project64.exe is "HLE GFX=1", when I, let's say, want to test an LLE plugin, so I uncheck it, and the setting "HLE GFX=0" gets written in the config, and in the UI it gets bolded. Then I'm done with with testing, and I change it back to "HLE GFX=1", which is what is coded in Project64.exe, and the setting in the cfg changes to "HLE GFX=1", which matches with the hardcoded, default settings found in the exe. But the thing is, ANYTHING found in the CFG is considered by the emulator as an override, so even if "HLE GFX=1" matches with the default, it is considered an override, so it gets bolded. To avoid this situation what I suggest is just to implement a check in which the emulator compares the current global settings with what's coded in the exe, so when "HLE GFX=1" is attempted to be written in the cfg, the emulator notices it is already matching with what's in the executable, so it just deletes the "HLE GFX=0" entry written in the cfg due to me changing the setting before.

There is already a cross-check between the RDB and the CFG file, so this principle works under the per-game settings panel. So if the RDB entry for DK64 says "RDRAM Size=8" but I want to see if it boots without the expansion pack, "RDRAM Size=4" gets written to the CFG, and the emulator reads the CFG and knows I'm overriding a setting. When I notice that the game doesn't boot with 4MB of RDRAM, I change it back to "RDRAM Size=8", the emulator checks that there's already an entry in the RDB for DK64 that says "RDRAM Size=8", so it deletes its entry from the config, otherwise, if the entry "RDRAM Size=8" were found in the RDB, it would be considered an override, so it would get bolded.

Even though that's effin stupid, because the point of bolding the setting in the first place is just to say that a setting no longer matches the default. If it didn't happen this way, any setting that I Happen to touch would be bolded forever until I press the "Reset Page/All" button 😛

**Comment by project64** [comment]
void CSettingTypeApplication::Save(int /*Index*/, uint32_t Value) 
has a check that will delete the setting if it is a default.

void CSettingTypeApplication::Save(int /*Index*/, bool Value)
does not

so just need to add the code to CSettingTypeApplication::Save to delete the value instead of setting it.

**Comment by cxd4** [comment]
tl;dr

Delete the Win32 GUI and everything will work out.  No more bold text.

**Comment by project64** [comment]
does https://github.com/project64/project64/commit/f8e49dfbb88b9053381662f8d82585e98c259761 fix it?

**Comment by Frank-74** [comment]
It works for the debugger.  It was already bolded in my settings, but unchecked.  I ran the new exe with this fix, it was still bolded.  I checked debugger, clicked apply/ok, then opened settings again and unchecked it clicked apply.  It unbolded ok.

I haven't checked the .cfg to see if the debugger section is gone.  I imagine it is though.  This should also fix the problem I had of leaving logging options enabled and disabling the debugger, the logging kept happening with the debugger disabled, because of the settings left in the .cfg.

**Comment by project64** [comment]
I probably should add a watch on the debug setting, when it is disabled, reset some of the debug settings back to there default.

**Comment by Frank-74** [comment]
I edited while you posted a reply.

It would be ok if the debugger remembered what you had set in the .cfg.  But the options need to be wrapped in a 'if debugger enabled'.

**Comment by oddMLan** [comment]
@Frank-74 , unrelated to this issue but I see that the emulator is creating an entry for every game in the CFG, and it seems to be due to the AI Count Per Byte being dynamically generated. Personally I don't think it's necessary to store this value, since it's generated on the fly anyways. It's redundant and just cluttering the CFG file. Also that option remains bolded in the Config panel due to that. The CFG should store it only when AICount =! -1 (default on-the-fly generated value)

**Comment by Frank-74** [comment]
@oddMLan Are you using a build I did sometime ago, that put the auto calculated AI value into the AI count per byte box after running the game?  I only made that hack, to see what the auto value was, so I could fine tune it for the DK rap timing.

The current git source only writes .cfg settings that are not in the rdb.  So it only writes to AI Count if you write a new value.  I don't have any AICount=***, in my current .cfg.

Here's my most recent build, https://dl.dropboxusercontent.com/u/9498358/n64/Project64.zip
The only difference from the current git source, I added hotkeys back to Glide64, CTRL+V, CTRL+B for buffer read always and motion blur, BACKSPACE to select filtering mode.  And added Azimer XA2 with a few changes.  Works well with the current default settings.  If a game has crackling, enabled sync using audio in PJ64.  That setting is same as turning off framelimiter with sync on in the audio plugin.

**Comment by project64** [comment]
@Frank-74  I been thinking of removing the hot keys out of glide64 .. how useful are they to you?

**Comment by oddMLan** [comment]
It works now, but changing settings in the Per-game settings panel is triggering a Fatal error now.

- Select any game from the list
- Uncheck or check "32-bit engine"
- Fatal error (and it doesn't save)

**Comment by Frank-74** [comment]
>  I been thinking of removing the hot keys out of glide64 .. how useful are they to you?

@project64 Toggle for FB Read Always is useful, and I do like to switch filtering modes to see the differences.

I'm also getting fatal error when switching any checkbox options on the per game settings except for Fixed Audio and Sync using audio.  Caused by the https://github.com/project64/project64/commit/f8e49dfbb88b9053381662f8d82585e98c259761 commit.

**Comment by project64** [comment]
https://github.com/project64/project64/commit/ec5c860a60f19d09f51b5a1b1257ccfcd9bc8b51 should fix the new crash

**Comment by project64** [comment]
I am trying to get angle working, but I would like to clean up glide64 ,,,, 

if I am using hotkeys I would want them to go though the core .. I dislike the plugin monitoring key strokes

**Comment by Frank-74** [comment]
New crash is fixed. @oddMLan I updated the link above if you want to grab it.

@project64 Anything to get angle working!  I'd happily live without hotkeys with angle.

**Comment by oddMLan** [comment]
Got it. The Fatal error has been fixed. 😁 

>Also if you change plugins (eg, select PJGlide64 then click "Apply", it gets bolded, then change back to Jabo's, and click "Apply" it doesn't get unbolded)

but this is still happening

**Comment by Frank-74** [comment]
** Use System Plugin ** option gets unbolded here.  The only thing that doesn't get unbolded is the Plugins on the left, unless I click reset page.

**Comment by Frank-74** [comment]
@oddMLan Updated the link above again, new Azimer XA2 that saves and remembers the sync settings, and made it the default audio plugin in Project64.  HLE Audio now default as well.

**Comment by cxd4** [comment]
I don't think it's necessary to have keyboard shortcuts etched inside the graphics plugin.

This dosn't really end up being more portable than menu options in the graphical config dialog, and is already a complex-enough subject of porting for something as simple as an input plugin.

**Comment by project64** [comment]
https://github.com/project64/project64/commit/bf0ba50dc494f21b91a353c3ab9a926a5699d0d8 should fix for plugins .. so this should be this bug fixed

**Comment by oddMLan** [comment]
Fixed. 
But it creates #1263 as a sideffect

**Comment by Frank-74** [comment]
> Fixed.
But commit bf0ba50 creates #1263 as a sideffect

#1263 fixed with #1267.
